### PR TITLE
Add default IA3 target modules for Mixtral

### PR DESCRIPTION
**Pull request by arnavgarg1**
Here's the mixtral model architecture with my proposed IA3 target module mapping:

```
PeftModelForCausalLM(                                                                                                                                       
  (base_model): IA3Model(                                                                                                                                   
    (model): MixtralForCausalLM(                                                                                                                            
      (model): MixtralModel(                                                                                                                                
        (embed_tokens): Embedding(32000, 4096)                                                                                                              
        (layers): ModuleList(                                                                                                                               
          (0-13): 14 x MixtralDecoderLayer(                                                                                                                 
            (self_attn): MixtralAttention(                                                                                                                  
              (q_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)                                                                         
              (k_proj): ia3.Linear4bit(                                                                                                                     
                (base_layer): Linear4bit(in_features=4096, out_features=1024, bias=False)                                                                   
                (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1024x1 (cuda:0)])                                  
              )                                                                                                                                             
              (v_proj): ia3.Linear4bit(                                                                                                                     
                (base_layer): Linear4bit(in_features=4096, out_features=1024, bias=False)                                                                   
                (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1024x1 (cuda:0)])                                  
              )                                                                                                                                             
              (o_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)                                                                         
              (rotary_emb): MixtralRotaryEmbedding()                                                                                                        
            )                                                                                                                                               
            (block_sparse_moe): MixtralSparseMoeBlock(                                                                                                      
              (gate): Linear4bit(in_features=4096, out_features=8, bias=False)                                                                              
              (experts): ModuleList(                                                                                                                        
                (0-7): 8 x MixtralBLockSparseTop2MLP(                                                                                                       
                  (w1): ia3.Linear4bit(                                                                                                                     
                    (base_layer): Linear4bit(in_features=4096, out_features=14336, bias=False)                                                              
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x4096 (cuda:0)])                              
                  )                                                                                                                                         
                  (w2): ia3.Linear4bit(                                                                                                                     
                    (base_layer): Linear4bit(in_features=14336, out_features=4096, bias=False)                                                              
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x14336 (cuda:0)])                             
                  )                                                                                                                                         
                  (w3): ia3.Linear4bit(                                                                                                                     
                    (base_layer): Linear4bit(in_features=4096, out_features=14336, bias=False)                                                              
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x4096 (cuda:0)])                              
                  )                                                                                                                                         
                  (act_fn): SiLU()                                                                                                                          
                )                                                                                                                                           
              )                                                                                                                                             
            )                                                                                                                                               
            (input_layernorm): MixtralRMSNorm()                                                                                                             
            (post_attention_layernorm): MixtralRMSNorm()
          )
          (14-31): 18 x MixtralDecoderLayer(
            (self_attn): MixtralAttention(
              (q_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)
              (k_proj): ia3.Linear4bit( 
                (base_layer): Linear4bit(in_features=4096, out_features=1024, bias=False)
                (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1024x1 (cuda:1)])
              )
              (v_proj): ia3.Linear4bit( 
                (base_layer): Linear4bit(in_features=4096, out_features=1024, bias=False)
                (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1024x1 (cuda:1)])
              )
              (o_proj): Linear4bit(in_features=4096, out_features=4096, bias=False)
              (rotary_emb): MixtralRotaryEmbedding()
            )
            (block_sparse_moe): MixtralSparseMoeBlock(
              (gate): Linear4bit(in_features=4096, out_features=8, bias=False) 
              (experts): ModuleList(
                (0-7): 8 x MixtralBLockSparseTop2MLP(
                  (w1): ia3.Linear4bit( 
                    (base_layer): Linear4bit(in_features=4096, out_features=14336, bias=False)
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x4096 (cuda:1)])
                  )
                  (w2): ia3.Linear4bit( 
                    (base_layer): Linear4bit(in_features=14336, out_features=4096, bias=False)
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x14336 (cuda:1)])
                  )
                  (w3): ia3.Linear4bit( 
                    (base_layer): Linear4bit(in_features=4096, out_features=14336, bias=False)
                    (ia3_l): ParameterDict(  (default): Parameter containing: [torch.cuda.HalfTensor of size 1x4096 (cuda:1)])
                  )
                  (act_fn): SiLU()
                )
              )
            )
            (input_layernorm): MixtralRMSNorm()
            (post_attention_layernorm): MixtralRMSNorm()
          )
        )
        (norm): MixtralRMSNorm()
      )
      (lm_head): Linear(in_features=4096, out_features=32000, bias=False)
    )
  )
)
```

These are the number of trainable parameters:

```
trainable params: 11,665,408 || all params: 46,714,458,112 || trainable%: 0.024971729249286513
```

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Hi @arnavgarg1 Thanks for the contribution! 
I did not noticed your PR until now .. I made https://github.com/huggingface/peft/pull/1380 few days ago that adds mixtral in the LoRA mapping. Would you be happy to convert this PR to a PR that adds mixtral to IA3 mapping instead?

**Comment by arnavgarg1** [comment]
@younesbelkada Yes!

**Review by pacman100 — COMMENTED** [review]
Hello @arnavgarg1, as Younes mentioned, please update the PR to add target modules for Mixtral when using IA3.

**Comment by arnavgarg1** [comment]
@pacman100 @younesbelkada Just updated with IA3 instead! I'm also going to add a separate PR for Phi with IA3 right now.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot!

**Comment by arnavgarg1** [comment]
No problem!

File: src/peft/utils/constants.py [hunk 1]
```diff
@@ -93,6 +93,7 @@ def starcoder_model_postprocess_past_key_value(past_key_values):
     "gpt_bigcode": ["c_attn", "mlp.c_proj"],
     "llama": ["k_proj", "v_proj", "down_proj"],
     "mistral": ["k_proj", "v_proj", "down_proj"],
+    "mixtral": ["k_proj", "v_proj", "w1", "w2", "w3"],
```
Comment (arnavgarg1) [inline src/peft/utils/constants.py:96]: @pacman100 Would this be `k_proj`, `v_proj` and `w2`?

**Comment by arnavgarg1** [comment]
Wanted to see what's left here as next steps @pacman100 @younesbelkada

**Review by younesbelkada — APPROVED** [review]
LGTM ! wdyt @pacman100 ?

File: src/peft/utils/constants.py [hunk 2]
```diff
@@ -115,6 +116,7 @@ def starcoder_model_postprocess_past_key_value(past_key_values):
     "gpt_bigcode": ["mlp.c_proj"],
     "llama": ["down_proj"],
     "mistral": ["down_proj"],
+    "mixtral": ["w1", "w2", "w3"],
```
Comment (pacman100) [inline src/peft/utils/constants.py:119]: As discussed above, it should only have `w2`

**Review by pacman100 — COMMENTED** [review]
Hello @arnavgarg1, thank you for adding IA3 target modules for Mixtral! Please see the comment on and post addressing that we can merge this.

**Comment by arnavgarg1** [comment]
Thanks @pacman100 ! Just updated

**Comment by arnavgarg1** [comment]
Is it good to merge?

**Comment by pacman100** [comment]
Thank you @arnavgarg1! ✨

**Review by pacman100 — APPROVED** [review]

**Comment by arnavgarg1** [comment]
Thanks!

**Pull request merged**